src/Algebra/Graph/AdjacencyIntMap/Algorithm.hs

```diff
@@ -91,17 +96,13 @@ bfsForest g = bfsForestFrom' (vertexList g) g
 -- 
 -- @
 bfsForestFrom :: [Int] -> AdjacencyIntMap -> Forest Int
-bfsForestFrom vs g = bfsForestFrom' [ v | v <- vs, hasVertex v g ] g
+bfsForestFrom vs g = evalState (bfsForestFrom' g vs) IntSet.empty
 
-bfsForestFrom' :: [Int] -> AdjacencyIntMap -> Forest Int
-bfsForestFrom' vs g = evalState (explore vs) IntSet.empty where
-  explore (v:vs) = discovered v >>= \case
-    True -> (:) <$> unfoldTreeM_BF walk v <*> explore vs
-    False -> explore vs
-  explore [] = return []
+bfsForestFrom' :: AdjacencyIntMap -> [Int] -> State IntSet.IntSet (Forest Int)
+bfsForestFrom' g = unfoldForestM_BF walk <=< filterM discovered where
   walk v = (v,) <$> adjacentM v
   adjacentM v = filterM discovered $ IntSet.toList (postIntSet v g)
-  discovered v = do new <- gets (not . IntSet.member v)
+  discovered v = do new <- gets ((&& hasVertex v g) . not . IntSet.member v)
```

Comment:
Checking here probably isn't the right move. The old way became awkward with the new semantics, but this seems to have impacted performance slightly

Reply:
Wait, what's awkward about keeping `[ v | v <- vs, hasVertex v g ]` at the top level as before?

Reply:
Nevermind! not sure what I was thinking